btk/catalog.py

```diff
@@ -6,66 +6,89 @@
 
 
 class Catalog(ABC):
-    def __init__(self, catalog_file, selection_function=lambda x: x, verbose=False):
+    def __init__(self, catalog, verbose=False):
         """Returns astropy table with catalog name from input class.
 
         Args:
-            catalog_file: File path of CatSim-like catalog or galsim COSMOS catalog to
-                         draw galaxies from.
-            selection_function: Selection cuts (if input) to place on input catalog.
+            catalog : CatSim-like catalog or galsim COSMOS catalog to draw galaxies from.
             verbose: Whether to print information related to loading catalog.
 
         Attributes:
-            self.cat (varies): Catalog object of varying type. Might be necessary to
-                                ultimately generate the images (e.g. COSMOS Catalog)
-            self.table (`astropy.table`): CatSim-like catalog with a selection
-                                           criteria applied if provided.
+            self.table (`astropy.table`): CatSim-like catalog with selection criteria applied 
+                and recorded in the `_selection_functions` list.
         """
-        self.cat = self.get_catalog(catalog_file)
-        self.table = selection_function(self.get_table())
+        self._raw_catalog = catalog
         self.verbose = verbose
+        self.table = self._prepare_table()
+        self._selection_functions = []
 
         if self.verbose:
             print("Catalog loaded")
 
     @abstractmethod
-    def get_catalog(self, catalog_file):
+    @classmethod
+    def from_file(cls, catalog_file, verbose):
+        """Catalog constructor from input file"""
         pass
 
     @abstractmethod
-    def get_table(self):
+    def _prepare_table(self):
+        """Operations to standardize the catalog table"""
         pass
 
     @property
     def name(self):
         return self.__class__.__name__
 
+    def get_raw_catalog(self):
+        return self._raw_catalog
+
+    def apply_selection_function(self, selection_function):
```

Comment:
I am still perplex about the selection_function ; shouldn't it be applied to the whole catalog in some way, and then the table gets extracted ? I do not really see an use case, aside some really basic things like slicing the catalog, which could be done in the sampling function anyway (whose job is basically to modify the table to represent the blend)

Reply:
The way I designed it, the `table` attribute is a sliced view of the raw catalog. Therefore, you can always come back to the original catalog. Whereas if you apply cuts to the catalog itself the changes are lost.

The other confusion you mention apparently comes from vocabulary when we take a _sampling function_ for a _selection function_. The way I understand it is that a sampling function is a carefully chosen set of selection functions. 
Do I understand correctly ?

Reply:
When we talk about sampling function it refers to the SamplingFunction object (which used to be a function but now is a callable object). It is where the details of the blends (for instance the positions of the galaxies etc) are defined, by modifiying the astropy table. As of now, I do not use a selection function and do everything in the sampling function, including things like mag cut which only restrict the catalog ; which is why I wonder if the selection function is really needed.

Reply:
Well, then that is exactly what I meant by "a set of selection functions" 😄 .
But I agree, the question is, do we need the possibility to do that at the Catalog level ?
For me it makes sense because there are some selections that can be relevant based on the knowledge of the catalog but independent of the scope filled by the SamplingFunction step.

Reply:
I agree with @thuiop that everything that is done in a "selection function" can be done in a "sampling function". But I think there are also situations where doing things at a catalog level initially might make the procedure cleaner later. 

Say that a user wants blends created from random galaxies that satisfy "i_ab < 25.3", I think in this case it is overkill for the user to have to create their own sampling function class. It would be much easier if they just specify their cuts as a selection function and use a already implemented sampling function that returns random entries from a given catalog (after cuts have been applied)